### PR TITLE
feat: flash close short and buy long

### PR DESCRIPTION
**Pull request by haythem96**
# feat: flash close short and buy long

## Description

This PR add the feature of closing a short position and buying WPowerPerp with the remaining funds.
The `flashWBurnBuyLong()` will:
- flashswap WPowerPerp token
- close the short position
- Remove collateral and repay flashswap
- Long WPowerPerp token using remaining ETH

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

### CODE REVIEW

File: packages/hardhat/contracts/periphery/ControllerHelper.sol [hunk 1]
```diff
@@ -109,6 +137,46 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
         emit FlashWMint(msg.sender, _vaultId, _wPowerPerpAmount, amountToFlashswap, _collateralAmount);
```
Comment (aleone) [inline packages/hardhat/contracts/periphery/ControllerHelper.sol:137]: I wonder if this should be passed in from f/e or just grabbed from the vault as the shortAmount

File: packages/hardhat/contracts/periphery/ControllerHelper.sol [hunk 2]
```diff
@@ -109,6 +137,46 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
...
     }
 
+    /**
+     * @notice flash close position and buy long squeeth
```
Comment (aleone) [inline packages/hardhat/contracts/periphery/ControllerHelper.sol:141]: I think this can be simplified with just 1 flashswap call

- totalToBuy = _wPowerPerpAmount + _minToReceive
- maxToPay = msg.value + vault.collateralAmount - _collateralToWithdraw

do everything in 1 flashswap callback (burn + withdraw collateral)

**Review by aleone — CHANGES_REQUESTED** [review]
let me know what you think about this refactor

File: packages/hardhat/contracts/periphery/ControllerHelper.sol [hunk 3]
```diff
@@ -109,6 +137,46 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
...
+     * @param _vaultId vault ID
+     * @param _wPowerPerpAmount amount of WPowerPerp to burn
+     * @param _collateralToWithdraw amount of collateral to withdraw
+     * @param _minToReceive minimum amount of long WPowerPerp to receive
+     */
+    function flashWBurnBuyLong(
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmount,
+        uint256 _collateralToWithdraw,
+        uint256 _minToReceive
```
Comment (aleone) [inline packages/hardhat/contracts/periphery/ControllerHelper.sol:151]: think this should be an exactOut swap (so wPowerPerpAmount to burn + wPowerPerpAmount to go long)

think should have a maxToPay param

File: packages/hardhat/contracts/periphery/ControllerHelper.sol [hunk 4]
```diff
@@ -109,6 +137,46 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
...
     }
 
+    /**
+     * @notice flash close position and buy long squeeth
+     * @param _vaultId vault ID
+     * @param _wPowerPerpAmount amount of WPowerPerp to burn
+     * @param _collateralToWithdraw amount of collateral to withdraw
+     * @param _minToReceive minimum amount of long WPowerPerp to receive
+     */
+    function flashWBurnBuyLong(
```
Comment (aleone) [inline packages/hardhat/contracts/periphery/ControllerHelper.sol:147]: maybe need to tweak names as well:

(if changed based on my comments) this is basically a function that can:
1. buyback and fully close a short
2. go long with any amount >= 0 wPowerPerp

So maybe fullCloseAndLong, closeShortAndBuy?

File: packages/hardhat/contracts/periphery/ControllerHelper.sol [hunk 5]
```diff
@@ -109,6 +137,37 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
...
+        uint256 _collateralToBuyWith
+    ) external payable {
+        require(_collateralToBuyWith <= _collateralToWithdraw.add(msg.value), "Not enough collateral");
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            IUniswapV3Pool(wPowerPerpPool).fee(),
+            _wPowerPerpAmountToBurn.add(_wPowerPerpAmountToBuy),
+            _collateralToBuyWith.add(msg.value),
```
Comment (haythem96) [inline packages/hardhat/contracts/periphery/ControllerHelper.sol:163]: We need to set here different slippage limit right ? @aleone

**Review by aleone — CHANGES_REQUESTED** [review]

**Review by alpinechicken — APPROVED** [review]
lgtm

File: packages/hardhat/test/integration-tests/periphery/controller-helper.ts [hunk 6]
```diff
@@ -146,5 +147,24 @@ describe("Controller helper integration test", function () {
...
+    it("flash close short position and buy long", async () => {
+      const vaultId = (await shortSqueeth.nextId()).sub(1);
+      await controller.connect(depositor).updateOperator(vaultId, controllerHelper.address)
+
+      const vaultBefore = await controller.vaults(vaultId)
+      const longBalanceBefore = await wSqueeth.balanceOf(depositor.address)
+      const squeethPrice = await oracle.getTwap(wSqueethPool.address, wSqueeth.address, weth.address, 420, true)
+      const squeethToBuy = vaultBefore.collateralAmount.div(squeethPrice)
+
+      await controllerHelper.connect(depositor).flashswapWBurnBuyLong(vaultId, vaultBefore.shortAmount, squeethToBuy, vaultBefore.collateralAmount, vaultBefore.collateralAmount);
```
Comment (aleone) [inline packages/hardhat/test/integration-tests/periphery/controller-helper.ts:160]: I think we should do something similar with slippage that is done in flash-mint to calculate the maxToPay param.

I also think for both of them we need to check that the amount returned to the user is what we expect.  (But believe you want to do this in a separate PR)

**Review by aleone — APPROVED** [review]
lgtm, but do think we need to improve the tests with slippage calculation as well as expected proceeds and checking that the user receives the correct amount back
